snapshots/devmapper/snapshotter.go

```diff
@@ -0,0 +1,425 @@
+	log.L.Debug("close")
+
+	var result *multierror.Error
+	s.closeOnce.Do(func() {
+		for _, fn := range s.cleanupFn {
+			if err := fn(); err != nil {
+				result = multierror.Append(result, err)
+			}
+		}
+	})
+
+	return result.ErrorOrNil()
+}
+
+func (s *Snapshotter) createSnapshot(ctx context.Context, kind snapshots.Kind, key, parent string, opts ...snapshots.Opt) ([]mount.Mount, error) {
+	snap, err := storage.CreateSnapshot(ctx, kind, key, parent, opts...)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(snap.ParentIDs) == 0 {
+		deviceName := s.getDeviceName(snap.ID)
+		log.G(ctx).Debugf("creating new thin device '%s'", deviceName)
+
+		err := s.pool.CreateThinDevice(ctx, deviceName, s.config.BaseImageSizeBytes)
+		if err != nil {
+			log.G(ctx).WithError(err).Errorf("failed to create thin device for snapshot %s", snap.ID)
+			return nil, err
+		}
+
+		if err := s.mkfs(ctx, deviceName); err != nil {
+			return nil, err
+		}
+	} else {
+		parentDeviceName := s.getDeviceName(snap.ParentIDs[0])
+		snapDeviceName := s.getDeviceName(snap.ID)
+		log.G(ctx).Debugf("creating snapshot device '%s' from '%s'", snapDeviceName, parentDeviceName)
+
+		err := s.pool.CreateSnapshotDevice(ctx, parentDeviceName, snapDeviceName, s.config.BaseImageSizeBytes)
+		if err != nil {
+			log.G(ctx).WithError(err).Errorf("failed to create snapshot device from parent %s", parentDeviceName)
+			return nil, err
+		}
+	}
+
+	mounts := s.buildMounts(snap)
+
+	// Remove default directories not expected by the container image
+	_ = mount.WithTempMount(ctx, mounts, func(root string) error {
+		return os.Remove(filepath.Join(root, "lost+found"))
+	})
+
+	return mounts, nil
+}
+
+// mkfs creates ext4 filesystem on the given devmapper device
+func (s *Snapshotter) mkfs(ctx context.Context, deviceName string) error {
+	args := []string{
+		"-E",
+		// We don't want any zeroing in advance when running mkfs on thin devices (see "man mkfs.ext4")
+		"nodiscard,lazy_itable_init=0,lazy_journal_init=0",
+		dmsetup.GetFullDevicePath(deviceName),
+	}
+
+	log.G(ctx).Debugf("mkfs.ext4 %s", strings.Join(args, " "))
+	output, err := exec.Command("mkfs.ext4", args...).CombinedOutput()
+	if err != nil {
+		log.G(ctx).WithError(err).Errorf("failed to write fs:\n%s", string(output))
+		return err
+	}
+
+	log.G(ctx).Debugf("mkfs:\n%s", string(output))
+	return nil
+}
+
+func (s *Snapshotter) getDeviceName(snapID string) string {
+	// Add pool name as prefix to avoid collisions with devices from other pools
+	return fmt.Sprintf("%s-snap-%s", s.config.PoolName, snapID)
+}
+
+func (s *Snapshotter) getDevicePath(snap storage.Snapshot) string {
+	name := s.getDeviceName(snap.ID)
+	return dmsetup.GetFullDevicePath(name)
+}
+
+func (s *Snapshotter) buildMounts(snap storage.Snapshot) []mount.Mount {
+	var options []string
+
+	if snap.Kind != snapshots.KindActive {
+		options = append(options, "ro")
+	}
+
+	mounts := []mount.Mount{
+		{
+			Source:  s.getDevicePath(snap),
```

Comment:
Devicemapper as in docker would create a 'rootfs' directory at block device and store image files under it. Should it respect that (at least, kata-runtime followed it)?

Reply:
If I get your question correctly: `getDevicePath` leads to `/dev/mapper/thin-device-name`, which is standard location for devmapper block devices. What exactly is going to be stored on these thin devices depends on containerd/runtime (so snapshotter manages block devices and returns mounts, containerd/runtime mounts device/writes data to it)

Reply:
I mean every filesystem hierarchy with the dm device you created should be `/roots/${imagefs}` instead of `/${imagefs}`

Reply:
@linxiulei The containerd maintainers can correct me, but my understanding is that the `/rootfs` prefix is a Docker behavior rather than an expectation of containerd snapshotters.  Unless we have a compelling reason for introducing another directory into the hierarchy, I'd prefer not to add it here.

Reply:
Yes, it’s what I suggest and considering some runtimes already implemented it that way. At least it’s worth to bring it up